library/src/main/java/org/apache/apex/malhar/lib/window/impl/AbstractWindowedOperator.java

```diff
@@ -0,0 +1,490 @@
+            windowStateMap.put(window, new WindowState());
+          }
+          windows.add(window);
+        }
+      } else if (windowOption instanceof WindowOption.SessionWindows) {
+        assignSessionWindows(windows, timestamp, inputTuple);
+      }
+    }
+  }
+
+  protected void assignSessionWindows(List<Window> windows, long timestamp, Tuple<InputT> inputTuple)
+  {
+    throw new UnsupportedOperationException();
+  }
+
+  /**
+   * Returns the list of windows TimeWindows for the given timestamp.
+   * If we are doing sliding windows, this will return multiple windows. Otherwise, only one window will be returned.
+   * Note that this method does not apply to SessionWindows.
+   *
+   * @param timestamp
+   * @return
+   */
+  private List<Window.TimeWindow> getTimeWindowsForTimestamp(long timestamp)
+  {
+    List<Window.TimeWindow> windows = new ArrayList<>();
+    if (windowOption instanceof WindowOption.TimeWindows) {
+      long durationMillis = ((WindowOption.TimeWindows)windowOption).getDuration().getMillis();
+      long beginTimestamp = timestamp - timestamp % durationMillis;
+      windows.add(new Window.TimeWindow(beginTimestamp, durationMillis));
+      if (windowOption instanceof WindowOption.SlidingTimeWindows) {
+        long slideBy = ((WindowOption.SlidingTimeWindows)windowOption).getSlideByDuration().getMillis();
+        // add the sliding windows front and back
+        // Note: this messes up the order of the window and we might want to revisit this if the order of the windows
+        // matter
+        for (long slideBeginTimestamp = beginTimestamp - slideBy;
+            slideBeginTimestamp <= timestamp && timestamp < slideBeginTimestamp + durationMillis;
+            slideBeginTimestamp -= slideBy) {
+          windows.add(new Window.TimeWindow(slideBeginTimestamp, durationMillis));
+        }
+        for (long slideBeginTimestamp = beginTimestamp + slideBy;
+            slideBeginTimestamp <= timestamp && timestamp < slideBeginTimestamp + durationMillis;
+            slideBeginTimestamp += slideBy) {
+          windows.add(new Window.TimeWindow(slideBeginTimestamp, durationMillis));
+        }
+      }
+    } else {
+      throw new IllegalStateException("Unexpected WindowOption");
+    }
+    return windows;
+  }
+
+  @Override
+  public boolean isTooLate(long timestamp)
+  {
+    return allowedLatenessMillis < 0 ? false : (timestamp < currentWatermark - allowedLatenessMillis);
+  }
+
+  @Override
+  public void dropTuple(Tuple<InputT> input)
+  {
+    // do nothing
+    LOG.debug("Dropping late tuple {}", input);
+  }
+
+
+  @Override
+  public void processWatermark(ControlTuple.Watermark watermark)
+  {
+    currentWatermark = watermark.getTimestamp();
+    long horizon = currentWatermark - allowedLatenessMillis;
+    if (allowedLatenessMillis >= 0) {
+      // purge window that are too late to accept any more input
+      dataStorage.removeUpTo(horizon);
```

Comment:
Before removing from storage, should we check whether triggers are fired and data is send to downstream for window that is too late?